WordPress/gutenberg
# Patterns: edit source pattern in focus mode in post and site editors

## What?
In the `Synced patterns partial syncing` experiment locks pattern editing in the post editor to only child blocks that have had content connections set.

**N.B.** This PR changes the paradigm for editing synced patterns in the editor and moves the editing of a pattern's content that causes global changes to a focus mode view, similar to template editing.

## Why?
As part of the work on allowing [partial editing of synced patterns](https://github.com/WordPress/gutenberg/issues/53705) we want to disable full editing of the synced pattern entities within the block editor.

This will make it easier to indicate to users when they are editing just a section of the pattern for that instance of the pattern only, versus editing the global source pattern instance.

This would also fix https://github.com/WordPress/gutenberg/issues/54442 as the edit button only shows if the user has permission to update the pattern. It also fixes #32353

## How?

- Sets each of the patterns innerBlocks edit mode to `disabled` using `setBlockEditingMode` unless the block connected attributes are set in which case the block is set to `contentOnly`. 
- Adds an `Edit` button to pattern block toolbar to edit source pattern.

**Important note** - limiting editing to `contentOnly` in the post editor is 🤞 just the starting point. We are keeping it restricted to contentOnly to keep things simple while we set up the initial APIs for implementing partial syncing. The hope is that we can later extend it out to things like innerBlocks, limited block style attributes, etc. when/if the complexities around doing so are resolved.

## Testing Instructions
- Make sure the synced pattern experiment is enabled
- Add a synced pattern 
- Insert the pattern into the block editor and check that the child blocks can't be edited, and don't appear in the list view
- Check that the `Edit original` button in the block toolbar redirects to focus mode canvas. Make sure pattern can be edited and that back button on top menu bar works as expected. 
- Check that changes to pattern show in post editor once when you go back and that the pattern shows as needing to be saved when you click  the post save button
- Check that the `Edit original` button does not display if the user does not have edit permission for the given pattern
- Also add the synced pattern in the site editor in a template and make sure `Edit original` and back buttons works as expected
- Add a new synced pattern using the following block markup, add the block to a post and make sure the Heading can't be edited but the paragraphs can be edited as `contentOnly`. Also check that the `Content` panel appears in right inspector panel when block or nested content selected.
- Turn of the synced pattern experiment and check that the synced patterns still behave the same as they do in trunk

```

<h2 class="wp-block-heading">This heading must stay the same</h2>



<p>You can change this</p>



<p>And you really should change this</p>

```

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/a5ac3082-da0f-4377-bd56-4ba8c97ec768
